keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,795 @@
+	// The two should match, but recall that clients can craft any kind of request.
+	// Webhook implementations must handle mismatched rawSelector and requirements, but how to handle them is up to the webhook.
+	Requirements []FieldSelectorRequirement
+}
+
+type LabelSelectorAttributes struct {
+	// rawSelector is the serialization of a field selector that would be included in a List query parameter.
+	// Webhook implementations must handle malformed rawSelectors.
+	RawSelector string
+
+    // requirements is the parsed interpretation of the rawSelector.
+    // All requirements must be met for a resource instance to match the selector.
+    // The two should match, but recall that clients can craft any kind of request.
+    // Webhook implementations must handle mismatched rawSelector and requirements, but how to handle them is up to the webhook.
+	Requirements []metav1.LabelSelectorRequirement
+}
+
+type FieldSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators are ==, !=
+	// TODO do we want to represent it in this API as In, NotIn, Exists and DoesNotExist.
+	// TODO we can convert existing field selectors into those operators and would ease future changes for webhooks.
+	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`
+	// values is an array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
+	// the values array must be empty. This array is replaced during a strategic
+	// merge patch.
+	// +optional
+	// +listType=atomic
+	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
+}
+
+
+```
+
+Importantly, if old webhook authorizers do not honor these new field, they will assume the broadest possible access and fail closed.
+If old in-cluster authorization does not include field and label selectors, the kube-apiserver will assume the broadest possible access and fail closed.
+
+### User Stories (Optional)
+
+#### As a SAR client, I want to check a request with a field or label selector
+
+This type of usage probably finds the stringified serialization format used in the query parameters the
+most convenient format to build their request with.
+Providing the query parameter serialization format avoids the need for a client to grow a decently complex lexer/parser.
+
+#### As an authorization webhook author, I want to easily consume the field and label selectors
+
+This type of usage probably finds a serialized `[]Requirement` to be the most convenient way to consume the field and label selector.
+Providing the parsed value avoids the need for every consumer to grow a decently complex lexer/parser.
+
+### Notes/Constraints/Caveats (Optional)
```

Comment:
some random spots to not forget to update:

* authorization webhook matchConditions, which evaluates the v1 SubjectAccessReview that _would_ be sent to the webhook https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go#L197-L205
* v1 / v1beta1 SAR translation function https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L472-L485
* v1 SAR construction function https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L198
* cache size decision https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L440